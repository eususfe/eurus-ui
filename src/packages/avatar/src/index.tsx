import { defineComponent } from 'vue';
import './style.scss';

const EAvatarProps = {
  size: {
    type: Boolean,
    default: false,
  },
  color: {
    type: String,
    default: '#fff',
  },
  offline: {
    type: Boolean,
    default: false,
  },
  notice: {
    type: Boolean,
    default: false,
  },
};

export default defineComponent({
  name: 'EAvatar',
  props: EAvatarProps,
  setup(props, { slots, emit }) {

    return () => (
      <div class={['avatar', props.size, props.offline && 'offline', props.notice && 'notice']} >
        {slots.default && slots.default()}
    </div>

    );
  },
});
