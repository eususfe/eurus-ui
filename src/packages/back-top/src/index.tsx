import type { PropType } from 'vue';
import { Transition, defineComponent, onBeforeUnmount, onMounted, ref, watchEffect } from 'vue';
import './style.scss';
import { getPrefixCls } from '@/packages/_utils/global-config';
import { isDocument, scrollTop } from '@/packages/_utils/dom';
import { warn } from '@/packages/_utils/warn';

const EBackTopProps = {
	text: {
		type: String,
		default: '返回顶部',
	},
	target: {
		type: [String, Object] as PropType<HTMLElement | string>,
		default: 'body',
	},
	height: { default: 200 },
	bottom: { default: 30 },
	right: { default: 30 },
};

export default defineComponent({
	name: 'EBackTop',
	props: EBackTopProps,
	emits: ['click'],
	setup(props, { slots, emit }) {
		const prefixCls = getPrefixCls('back-top');

		const backTopVisible = ref(false);

		let scrollListenerRegistered: boolean;
		let scrollElement: HTMLElement | Document;

		const handleScroll = () => {
			backTopVisible.value = scrollTop(scrollElement) > props.height;
		};
		const scrollToTop = () => {
			(isDocument(scrollElement) ? document.documentElement : scrollElement).scrollTo({
				top: 0,
				behavior: 'smooth',
			});
		};
		const backTopClick = () => {
			const currentScroll = scrollTop(scrollElement);
			if (currentScroll > 0) {
				scrollToTop();
			}
			emit('click');
		};

		onMounted(() => {
			if (scrollListenerRegistered) {
				return;
			}
			scrollListenerRegistered = true;
			const { target } = props;
			const targetElement: HTMLElement | Document =
				typeof target === 'string' ? (document.querySelector(target) as HTMLElement) : target;
			if (__DEV__) {
				watchEffect(() => {
					if (!targetElement) {
						warn('back-top', 'Target is not found.');
					}
				});
			}
			scrollElement = targetElement === document.documentElement ? document : targetElement;
			scrollElement.addEventListener('scroll', handleScroll);
		});

		onBeforeUnmount(() => {
			scrollElement.removeEventListener('scroll', handleScroll);
		});

		return () => (
			<Transition name="fade-in">
				<div
					class={[prefixCls]}
					role="button"
					aria-hidden="true"
					onClick={backTopClick}
					style={{
						right: props.right + 'px',
						bottom: props.bottom + 'px',
						visibility: backTopVisible.value ? 'visible' : 'hidden',
					}}
				>
					{slots?.default ? (
						slots.default()
					) : (
						<span class={[prefixCls + '-text']} v-text={props.text}></span>
					)}
				</div>
			</Transition>
		);
	},
});
