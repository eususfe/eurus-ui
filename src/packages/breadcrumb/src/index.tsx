import type { PropType } from 'vue';
import { defineComponent } from 'vue';
import { getPrefixCls } from '@/packages/_utils/global-config';
import './style.scss';

export interface BreadcrumbItem {
  title: string;
  href?: string;
}

const EBreadcrumbProps = {
  data: {
    type: Array as PropType<BreadcrumbItem[]>,
    default: ()=>[],
  },
  separator: {
    type: String,
    default: '/',
  },
};

export default defineComponent({
  name: 'EBreadcrumb',
  props: EBreadcrumbProps,
  setup(props) {
    const prefixCls = getPrefixCls('breadcrumb');

    return () => (
      <div class={[prefixCls]}>
        {
          props.data.map((item, index) => {
            return [item.href
              ? <router-link
                key={index}
                class={{ [prefixCls + '-item']: true, 'last-child': props.data.length === index + 1 }}
                to={item.href}
                v-text={item.title}
              />
              : <span class={{ [prefixCls + '__item']: true }} key={`span${index}`} v-text={item.title}></span>,

            props.data.length > index + 1
            && <span
              key={index + 100}
              class="separator"
            >{props.separator}</span >
            ];
          })
        }
      </div>

    );
  },
});
