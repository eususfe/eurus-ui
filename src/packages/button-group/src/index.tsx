import type { PropType } from 'vue';
import { defineComponent, provide } from 'vue';
import './style.scss';
import { getPrefixCls } from '@/packages/_utils/global-config';
import { buttonGroupKey } from '@/packages/_utils/constants';
import type { Size } from '@/packages/_utils/size';

const EButtonGroupProps = {
  vertical: {
    type: Boolean,
    default: false,
  },
  size: {
    type: String as PropType<Size>,
    default: 'md',
  },
};

export default defineComponent({
  name: 'EButtonGroup',
  props: EButtonGroupProps,
  setup(props, { slots }) {
    const prefixCls = getPrefixCls('button-group');

    provide(buttonGroupKey, props);
    return () => <div class={[prefixCls, props.vertical && 'vertical']}>{slots?.default && slots.default()}</div>;
  },
});
