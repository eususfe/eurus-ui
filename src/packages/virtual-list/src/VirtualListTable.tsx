import { defineComponent } from 'vue';

export default defineComponent({
  props: {
    table: {
      type: Boolean,
      default: false
    }
  },
  setup(props, { slots }) {
    return () => (
      props.table ? <table>
        {slots?.prepend && slots.prepend()}
        <tbody>
          {slots?.default && slots.default()}
        </tbody>
        {slots?.append && slots.append()}
      </table>
        : <div>
          {slots?.prepend && slots.prepend()}
          {slots?.default && slots.default()}
          {slots?.append && slots.append()}
        </div>
    );
  }
});
