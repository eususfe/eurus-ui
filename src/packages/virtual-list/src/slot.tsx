import { defineComponent, ref } from 'vue';
import { useResizeChange } from './item';
import { SlotProps } from './props';

export default defineComponent({
  name: 'VirtualListSlot',
  props: SlotProps,
  emits: ['slotResize'],
  setup(props, { slots, emit }) {
    const rootRef = ref<HTMLElement | null>(null);
    useResizeChange(props, rootRef, emit);

    return () => {
      const { tag: Tag, uniqueKey } = props;

      return (
        <Tag ref={rootRef} key={uniqueKey}>
          {slots.default?.()}
        </Tag>
      );
    };
  },
});
