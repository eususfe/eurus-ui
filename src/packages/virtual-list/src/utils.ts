export class Accumulate<T> {
  arr: T[] = [];
  cache: number[] = [];
  constructor(arr: T[]) {
    this.arr = arr;
  }

  getValue(item: T): number {
    // @ts-expect-error
    return item;
  }

  sum(index: number) {
    if (this.cache[index] == null) {
      const prev = index > 0 ? this.sum(index - 1) : 0;
      this.cache[index] = this.getValue(this.arr[index]) + prev;
    }
    return this.cache[index];
  }
}

/**
 * If n greater than `max`, return `max`, else n.
 * 如果n大于max, 返回max, 否则n.
 * @param n
 * @param max
 * @returns
 */
export function notGreaterThan<T>(n: T, max: T) {
  return n < max ? n : max;
}

/**
 * get last of array
 * 返回数组末项
 * @param arr
 * @returns
 */
export function arrayLast<T>(arr: T[]) {
  return arr[arr.length - 1];
}
/**
 * like jquery $(el).css(), but only can read
 * @param el
 * @param name
 * @returns
 */
export function css(el: Element, name: keyof CSSStyleDeclaration ): string {
  const stl = getComputedStyle(el);
  return stl[name] as string;
}

// ## advanced

export type BinarySearchReturn<T> = {
  index: number;
  value: T;
  count: number;
  hit: boolean;
  greater?: boolean;
} | null;
export interface BinarySearchOptions {
  start?: number;
  end?: number;
  returnNearestIfNoHit?: boolean;
  maxTimes?: number;
}
/**
 * binarySearch, 二分查找
 * @param arr
 * @param callback return `mid - your_value` for ascending array
 * @param opt
 * @returns
 */
export function binarySearch<T>(
  arr: T[],
  callback: (mid: T, index: number, count: number) => number,
  opt: BinarySearchOptions = {}
): BinarySearchReturn<T> {

  let { start = 0, end = arr.length - 1 } = opt;
  const { returnNearestIfNoHit, maxTimes = 100 } = opt;
  let midNum;
  let mid;
  if (start == null) {
    start = 0;
    end = arr.length - 1;
  }
  let i = 0;
  let r = 0;
  while (start >= 0 && start <= end) {
    if (i >= maxTimes) {
      throw new Error(
        `binarySearch: loop times is over ${maxTimes}, you can increase the limit.`
      );
    }
    midNum = Math.floor((end - start) / 2 + start);
    mid = arr[midNum];
    const count = i + 1;
    r = callback(mid, midNum, count);
    if (r > 0) {
      end = midNum - 1;
    } else if (r < 0) {
      start = midNum + 1;
    } else {
      return { index: midNum as number, value: mid as T, count, hit: true };
    }
    i++;
  }
  return returnNearestIfNoHit
    ? {
        index: midNum as number,
        value: mid as T,
        count: i + 1,
        hit: false,
        greater: r > 0,
      }
    : null;
}

// source: http://youmightnotneedjquery.com/
export function hasClass(el: Element, className: string) {
  if (el.classList) {
    return el.classList.contains(className);
  } else {
    return new RegExp('(^| )' + className + '( |$)', 'gi').test(el.className);
  }
}
