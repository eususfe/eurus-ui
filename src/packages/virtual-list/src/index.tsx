import type {
  PropType
} from 'vue';
import {
  onUnmounted,
  renderSlot,
  ref,
  computed,
  onMounted,
  nextTick,
  defineComponent,
  reactive,
  watch,
  h
} from 'vue';
import { Accumulate, arrayLast, binarySearch, css, hasClass, notGreaterThan } from './utils';
import VirtualListTable from './VirtualListTable';

const EVirtualListProps = {
  items: Array as PropType<any[]>,
  disabled: Boolean,
  horizontal: Boolean,
  firstRender: { type: Number, default: 10 },
  buffer: { type: Number, default: 100 },
  itemKey: {
    type: [String, Function] as PropType<
      'index' | ((item: any, index: number) => any)
    >,
  },
  itemSize: {
    type: Function as PropType<
      (item: any, index: number) => number | null | undefined | void
    >,
  },
  table: Boolean,
};
export default defineComponent({
  name: 'EVirtualList',
  props: EVirtualListProps,
  setup(props, { slots, emit }) {
    const start = ref(0);
    const end = ref(props.firstRender - 1);
    const end2 = computed(() =>
      notGreaterThan(end.value, (props.items?.length || 1) - 1)
    );
    const avgSize = ref(0);
    const runtimeSizes = computed(() =>
      reactive<(number | null)[]>((props.items || []).map(() => null))
    );

    const sizes = computed(() =>
      (props.items || []).map((item, index) =>
        computed(() => {
          if (runtimeSizes.value[index] != null) {
            return runtimeSizes.value[index];
          }
          let r = props.itemSize?.(item, index);
          if (r == null) {
            r = avgSize.value;
          }
          return r;
        })
      )
    );
    const positions = computed(() =>
      (props.items || []).map((item, index) =>
        computed(() => {
          let r: number;
          if (index === 0) {
            r = 0;
          }

          const prevPosition = positions.value![index - 1] || ref(0);
          const prevSize = sizes.value![index - 1] || ref(0);
          r = prevPosition.value + prevSize.value!;

          return r;
        })
      )
    );
    function getPosition(index: number): number {
      return positions.value[index].value;
    }
    // not reactive
    const positionsAccumulate = ref<Accumulate<typeof sizes.value[0]>>();

    const startSize = computed(() => getPosition(start.value));
    const totalSize = computed(() =>
      positions.value.length > 0
        ? getPosition(positions.value.length - 1)
        + arrayLast(sizes.value).value!
        : 0
    );
    const endSize = computed(
      () =>
        totalSize.value
        - getPosition(end2.value)
        - sizes.value[end2.value].value!
    );
    const listStyle = computed(() =>
      !props.disabled ? { overflow: 'auto' } : {}
    );
    const listInnerStyle = computed(() => {
      const r: any = {
        display: 'flex',
      };
      if (!props.disabled) {
        if (!props.horizontal) {
          Object.assign(r, {
            'margin-top': startSize.value + 'px',
            'margin-bottom': endSize.value + 'px',
          });
        } else {
          Object.assign(r, {
            'margin-left': startSize.value + 'px',
            'margin-right': endSize.value + 'px',
            'width': totalSize.value - endSize.value - startSize.value + 'px',
          });
        }
      }
      r['flex-direction'] = !props.horizontal ? 'column' : 'row';
      if (props.table) {
        delete r.display;
        delete r['flex-direction'];
      }
      return r;
    });

    const resetPositionsAccumulate = () => {
      positionsAccumulate.value = new Accumulate(sizes.value);
      positionsAccumulate.value.getValue = item => item['value']!;
    };
    resetPositionsAccumulate();

    watch(() => props.items, update);
    watch(() => positions.value, resetPositionsAccumulate);
    const visibleItemsInfo = computed(() => {
      if (!props.items || props.disabled) {
        return;
      }
      const r: { item: typeof props.items[0]; index: number }[] = [];

      for (let index = start.value; index <= end2.value; index++) {
        const item = props.items[index];
        if (!item) {
          break;
        }
        r.push({ item, index });
      }
      return r;
    });
    const listElRef = ref<HTMLElement>();
    const listInnerRef = ref();
    onMounted(async () => {
      update();
      try {
        createResizeObserver();
      } catch {
        // ResizeObserver fallback
        await nextTick();
        update();
      }
    });

    let prevScroll: number;
    function onscroll() {
      const listEl = listElRef.value!;
      if (!listEl) {
        return;
      }
      const currentScroll = getScroll(listEl);
      if (
        prevScroll != null
        && props.buffer - Math.abs(currentScroll - prevScroll) >= 10
      ) {
        return;
      }
      prevScroll = currentScroll;
      update();
    }

    let executing = false;
    let waiting = false;
    async function update() {
      if (executing) {
        waiting = true;
        return;
      }
      if (!props.items || props.disabled) {
        return;
      }
      executing = true;
      const listEl = listElRef.value!;
      const listInner: HTMLElement = listInnerRef.value?.$el;
      if (!listEl || !listInner) {
        return;
      }

      if (!avgSize.value) {
        avgSize.value = getAvgSize();
      }
      start.value = getStart();
      end.value = getEnd();

      await nextTick();

      // updateRuntimeSize
      let updated;
      let vi0 = 0;
      const runtimeSizesTemp: Record<number, number> = {};
      const children: HTMLCollection = !props.table
        ? listInner.children
        : listInner.querySelector('tbody')!.children;
      for (const el of Array.from(children)) {
        const cssPosition = css(el, 'position');
        if (cssPosition && ['absolute', 'fixed'].includes(cssPosition)) {
          continue;
        }
        const size
          = css(el, 'display') !== 'none' ? getOuterSize(el) : 0;
        const vi = el.getAttribute('vt-index');
        const index = vi ? Number.parseInt(vi) : start.value + vi0;
        runtimeSizesTemp[index] = (runtimeSizesTemp[index] || 0) + size;
        vi0++;
      }
      for (const indexS of Object.keys(runtimeSizesTemp)) {
        const index = Number.parseInt(indexS);
        if (runtimeSizes.value[index] !== runtimeSizesTemp[index]) {
          runtimeSizes.value[index] = runtimeSizesTemp[index];
          updated = true;
        }
      }
      if (updated) {
        await nextTick();
      }
      // call wating task if existing
      executing = false;
      if (waiting) {
        waiting = false;
        update();
      }

      // functions
      function getStart() {
        const startPosition
          = getScroll(listEl) - getPaddingStart(listEl) - props.buffer;
        const r = binarySearch(
          positions.value,
          mid => mid.value - startPosition,
          { returnNearestIfNoHit: true }
        )!;
        return r.index;
      }
      function getEnd() {
        const endPosition
          = getScroll(listEl)
          - getPaddingStart(listEl)
          + getClientSize(listEl)
          + props.buffer;

        const r = binarySearch(
          positions.value,
          mid => mid.value - endPosition,
          { returnNearestIfNoHit: true }
        )!;
        return r.index;
      }
      function getAvgSize() {
        const maxSampleCount = 10;
        const sizeArr: number[] = [];
        const children = !props.table
          ? listInner.children
          : listInner.querySelector('tbody')!.children;
        for (const child of Array.from(children)) {
          const el = child;
          const style = getComputedStyle(el);
          if (['absolute', 'fixed'].includes(style.position)) {
            continue;
          }
          const outerSize = getOuterSize(el);
          sizeArr.push(outerSize);
          if (sizeArr.length >= maxSampleCount) {
            break;
          }
        }
        if (sizeArr.length === 0) {
          return 0;
        }
        return sizeArr.reduce((a, b) => a + b, 0) / sizeArr.length;
      }
    }
    function getClientSize(el: Element) {
      const style = getComputedStyle(el);
      let r = Number.parseFloat(!props.horizontal ? style.height : style.width);
      if (style.boxSizing === 'border-box') {
        if (!props.horizontal) {
          r
            = r
            - Number.parseFloat(style.borderTopWidth)
            - Number.parseFloat(style.borderBottomWidth);
        } else {
          r
            = r
            - Number.parseFloat(style.borderLeftWidth)
            - Number.parseFloat(style.borderRightWidth);
        }
      }
      return r;
    }
    function getOuterSize(el: Element) {
      let r = getClientSize(el);
      const style = getComputedStyle(el);
      if (!props.horizontal) {
        r
          += Number.parseFloat(style.borderTopWidth)
          + Number.parseFloat(style.borderBottomWidth)
          + Number.parseFloat(style.marginTop)
          + Number.parseFloat(style.marginBottom);
      } else {
        r
          += Number.parseFloat(style.borderLeftWidth)
          + Number.parseFloat(style.borderRightWidth)
          + Number.parseFloat(style.marginLeft)
          + Number.parseFloat(style.marginRight);
      }
      r = Number.isNaN(r) ? 0 : r;
      return r;
    }
    function getScroll(el: HTMLElement) {
      return !props.horizontal ? el.scrollTop : el.scrollLeft;
    }
    function getPaddingStart(el: HTMLElement) {
      const style = getComputedStyle(el);
      return !props.horizontal
        ? Number.parseFloat(style.paddingTop)
        : Number.parseFloat(style.paddingLeft);
    }
    // function getPaddingEnd(el: HTMLElement) {
    //   const style = getComputedStyle(el);
    //   return !props.horizontal
    //     ? parseFloat(style.paddingBottom)
    //     : parseFloat(style.paddingRight);
    // }

    function createResizeObserver() {
      const listEl = listElRef.value!;
      const resizeObserver = new ResizeObserver((entries) => {
        for (const entry of entries) {
          if (hasClass(entry.target, 'vtlist')) {
            update();
            break;
          }
        }
      });
      resizeObserver.observe(listEl);
    }

    function getItemKey(item: any, index: number) {
      if (props.itemKey) {
        if (typeof props.itemKey === 'string' && props.itemKey === 'index') {
          return index;
        } else if (typeof props.itemKey === 'function') {
          return props.itemKey(item, index);
        }
      }
    }

    onMounted(() => {
      listElRef.value?.addEventListener('scroll', onscroll, { passive: true });
    });
    onUnmounted(() => {
      listElRef.value?.removeEventListener('scroll', onscroll);
    });
    return () => (
      <div class="vtlist" ref={listElRef} style={listStyle.value} >

        {h(VirtualListTable, {
          class: 'vtlist-inner',
          ref: listInnerRef,
          style: listInnerStyle.value,
          table: props.table,
        }, {
          prepend: () => [
            slots.prepend?.(),
          ],
          append: () => [
            slots.append?.(),
          ],
          default: () => {
            return ((props.disabled ? props.items : visibleItemsInfo.value) || []).map((item, index) => {
              return renderSlot(slots, 'default', {
                key: getItemKey(item, index),
                item,
                index
              });
            });
          }
        })}

      </div>

    );
  },
});
