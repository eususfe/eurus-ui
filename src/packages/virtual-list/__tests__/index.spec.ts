/*
 * @Author: jiangxd
 * @Date: 2022-09-19 13:36:16
 * @LastEditTime: 2022-09-22 17:48:57
 * @LastEditors: jiangxd
 * @Description:
 * @FilePath: /eurus-ui/src/packages/virtual-list/__tests__/index.spec.ts
 */
import { mount } from '@vue/test-utils';
import { markRaw } from 'vue';
import EVirtualList from '..';
import { VirtualProps } from '../src/props';
import Item from './item';
import { getDatas } from './util';

describe('EVirtualList', () => {
  const Instance = mount(EVirtualList, {
    props: {
      dataSources: getDatas(1000),
      dataComponent: markRaw( Item),
      dataKey: 'id',
      direction: 'horizontal',
      wrapClass: '{ \'display\': \'flex\', \'flex-direction\': \'row\' }'

    }, attrs: {
      style: 'width: 300px; height: 300px; overflow: auto;',
      class: 'my-list'
    }
  });

  it('EVirtualList snapshot', () => {
    expect(Instance.html()).toMatchSnapshot();
    expect(Instance.find('.my-list').exists()).toBe(true);

  });

});
