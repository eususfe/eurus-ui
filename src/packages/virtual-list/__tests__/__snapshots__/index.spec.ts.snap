// Vitest Snapshot v1

exports[`EVirtualList > EVirtualList snapshot 1`] = `
"<div style=\\"width: 300px; height: 300px; overflow: auto;\\" class=\\"my-list\\">
  <!---->
  <div class=\\"{ &#x27;display&#x27;: &#x27;flex&#x27;, &#x27;flex-direction&#x27;: &#x27;row&#x27; }\\" style=\\"padding: 0px 48500px 0px 0px;\\">
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">0</span><span class=\\"source\\">67b3959</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">1</span><span class=\\"source\\">54c76f4</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">2</span><span class=\\"source\\">e03c8d5</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">3</span><span class=\\"source\\">3115ae5</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">4</span><span class=\\"source\\">0351a69</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">5</span><span class=\\"source\\">34c6894</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">6</span><span class=\\"source\\">61357b1</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">7</span><span class=\\"source\\">24df8eb</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">8</span><span class=\\"source\\">4d09994</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">9</span><span class=\\"source\\">1d5b3d1</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">10</span><span class=\\"source\\">03a8242</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">11</span><span class=\\"source\\">7841708</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">12</span><span class=\\"source\\">be8beca</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">13</span><span class=\\"source\\">601e4a5</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">14</span><span class=\\"source\\">962f3d5</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">15</span><span class=\\"source\\">c4e353a</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">16</span><span class=\\"source\\">4016566</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">17</span><span class=\\"source\\">bb2bdbd</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">18</span><span class=\\"source\\">eefa6a7</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">19</span><span class=\\"source\\">386bf9b</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">20</span><span class=\\"source\\">c7b0261</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">21</span><span class=\\"source\\">2a31e76</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">22</span><span class=\\"source\\">2002afa</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">23</span><span class=\\"source\\">85870af</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">24</span><span class=\\"source\\">cdcc77e</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">25</span><span class=\\"source\\">0027001</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">26</span><span class=\\"source\\">590e826</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">27</span><span class=\\"source\\">809e9af</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">28</span><span class=\\"source\\">aff3a6</span><span class=\\"other\\"><!----></span></div>
    </div>
    <div class=\\"\\">
      <div class=\\"inner\\"><span class=\\"index\\">29</span><span class=\\"source\\">d20541c</span><span class=\\"other\\"><!----></span></div>
    </div>
  </div>
  <!---->
  <div style=\\"width: 0px; height: 100%;\\"></div>
</div>"
`;
